Mathlib/Geometry/Manifold/IntegralCurve.lean

```diff
@@ -0,0 +1,297 @@
+  `γ : ℝ → M` is a local integral curve of `v` in an open interval of `t₀`. That is, there exists
+  `ε > 0` such that `γ t` is tangent to `v (γ t)` for all `t ∈ Ioo (t₀ - ε) (t₀ + ε)`. The value of
+  `γ` outside of this interval is irrelevant and considered junk. -/
+def IsIntegralCurveAt (γ : ℝ → M) (v : (x : M) → TangentSpace I x) (t : ℝ) :=
+  ∃ ε > (0 : ℝ), IsIntegralCurveOn γ v (Ioo (t - ε) (t + ε))
+
+/-- If `v : M → TM` is a vector field on `M`, `IsIntegralCurve γ v` means `γ : ℝ → M` is a global
+  integral curve of `v`. That is, `γ t` is tangent to `v (γ t)` for all `t : ℝ`. -/
+def IsIntegralCurve (γ : ℝ → M) (v : (x : M) → TangentSpace I x) :=
+  ∀ t : ℝ, HasMFDerivAt 𝓘(ℝ, ℝ) I γ t ((1 : ℝ →L[ℝ] ℝ).smulRight (v (γ t)))
+
+variable {γ : ℝ → M} {v : (x : M) → TangentSpace I x} {x₀ : M}
+  (hv : ContMDiffAt I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)) x₀)
+  {s : Set ℝ} {t₀ : ℝ}
+
+lemma IsIntegralCurve.isIntegralCurveOn (h : IsIntegralCurve γ v) (s) :
+  IsIntegralCurveOn γ v s := fun t _ => h t
+
+lemma isIntegralCurve_iff_isIntegralCurveOn : IsIntegralCurve γ v ↔ IsIntegralCurveOn γ v univ :=
+  ⟨fun h => h.isIntegralCurveOn _, fun h t => h t (mem_univ _)⟩
+
+lemma IsIntegralCurve.isIntegralCurveAt (h : IsIntegralCurve γ v) (t : ℝ) :
+    IsIntegralCurveAt γ v t :=
+  ⟨1, zero_lt_one, fun t _ => h t⟩
+
+lemma isIntegralCurve_iff_isIntegralCurveAt :
+    IsIntegralCurve γ v ↔ ∀ t : ℝ, IsIntegralCurveAt γ v t :=
+  ⟨fun h => h.isIntegralCurveAt, fun h t => by
+    obtain ⟨ε, hε, h⟩ := h t
+    exact h t (Real.ball_eq_Ioo _ _ ▸ Metric.mem_ball_self hε)⟩
+
+lemma IsIntegralCurveOn.mono (h : IsIntegralCurveOn γ v s) {s' : Set ℝ} (hs : s' ⊆ s) :
+    IsIntegralCurveOn γ v s' :=
+  fun t ht => h t (mem_of_mem_of_subset ht hs)
+
+lemma IsIntegralCurveOn.isIntegralCurveAt (h : IsIntegralCurveOn γ v s) {t : ℝ} (hs : s ∈ nhds t) :
+    IsIntegralCurveAt γ v t := by
+  rw [Metric.mem_nhds_iff] at hs
+  obtain ⟨ε, hε, hmem⟩ := hs
+  exact ⟨ε, hε, Real.ball_eq_Ioo _ _ ▸ h.mono hmem⟩
+
+lemma IsIntegralCurveAt.isIntegralCurveOn (h : ∀ t ∈ s, IsIntegralCurveAt γ v t) :
+    IsIntegralCurveOn γ v s := by
+  intros t ht
+  obtain ⟨ε, hε, h⟩ := h t ht
+  exact h t (Real.ball_eq_Ioo _ _ ▸ Metric.mem_ball_self hε)
+
+/-! ### Translation lemmas -/
+section Translation
+
+lemma IsIntegralCurveOn.comp_add (hγ : IsIntegralCurveOn γ v s) (dt : ℝ) :
+    IsIntegralCurveOn (γ ∘ (· + dt)) v { t | t + dt ∈ s } := by
+  intros t ht
+  rw [Function.comp_apply,
+    ← ContinuousLinearMap.comp_id (ContinuousLinearMap.smulRight 1 (v (γ (t + dt))))]
+  apply HasMFDerivAt.comp t (hγ (t + dt) ht)
+  refine ⟨(continuous_add_right _).continuousAt, ?_⟩
+  simp only [mfld_simps, hasFDerivWithinAt_univ]
+  exact HasFDerivAt.add_const (hasFDerivAt_id _) _
+
+lemma isIntegralCurveOn_comp_add {dt : ℝ} :
+    IsIntegralCurveOn γ v s ↔ IsIntegralCurveOn (γ ∘ (· + dt)) v { t | t + dt ∈ s } := by
```

Comment:
I think this spelling is slightly more readable: a "new idea" goes on one line (e.g., lemma name, hypothesis, conclusion, proof --- if shorter, they can of course be combined; in this case, as the conclusion doesn't fit on one line anyway, I'd rather put it separately). Again, this is a really minor change.